linux/LinuxProcess.c

```diff
@@ -288,13 +594,95 @@ void LinuxProcess_writeField(Process* this, RichString* str, ProcessField field)
       xSnprintf(buffer, n, "%5lu ", lp->ctxt_diff);
       break;
    case SECATTR: snprintf(buffer, n, "%-30s   ", lp->secattr ? lp->secattr : "?"); break;
+   case COMM: {
+      if ((Process_isUserlandThread(this) && this->settings->showThreadNames) || !lp->mergedCommand.str) {
+         Process_writeField(this, str, field);
+         return;
+      }
+      /* This code is from Process_writeField for COMM, but we invoke
+       * LinuxProcess_writeCommand to display
+       * /proc/pid/exe (or its basename)│/proc/pid/comm│/proc/pid/cmdline */
+      int baseattr = CRT_colors[PROCESS_BASENAME];
+      if (this->settings->highlightThreads && Process_isThread(this)) {
+         attr = CRT_colors[PROCESS_THREAD];
+         baseattr = CRT_colors[PROCESS_THREAD_BASENAME];
+      }
+      if (!this->settings->treeView || this->indent == 0) {
+         LinuxProcess_writeCommand(this, attr, baseattr, str);
+         return;
+      } else {
+         char* buf = buffer;
+         int maxIndent = 0;
+         bool lastItem = (this->indent < 0);
+         int indent = (this->indent < 0 ? -this->indent : this->indent);
+         int vertLen = strlen(CRT_treeStr[TREE_STR_VERT]);
+
+         for (int i = 0; i < 32; i++) {
+            if (indent & (1U << i)) {
+               maxIndent = i+1;
+            }
+         }
+         for (int i = 0; i < maxIndent - 1; i++) {
+            if (indent & (1 << i)) {
+               if (buf - buffer + (vertLen + 3) > n) {
+                  break;
+               }
+               buf = stpcpy(buf, CRT_treeStr[TREE_STR_VERT]);
+               buf = stpcpy(buf, "  ");
+            } else {
+               if (buf - buffer + 4 > n) {
+                  break;
+               }
+               buf = stpcpy(buf, "   ");
+            }
+         }
+         n -= (buf - buffer);
+         const char* draw = CRT_treeStr[lastItem ? (this->settings->direction == 1 ? TREE_STR_BEND : TREE_STR_TEND) : TREE_STR_RTEE];
+         xSnprintf(buf, n, "%s%s ", draw, this->showChildren ? CRT_treeStr[TREE_STR_SHUT] : CRT_treeStr[TREE_STR_OPEN] );
+         RichString_append(str, CRT_colors[PROCESS_TREE], buffer);
+         LinuxProcess_writeCommand(this, attr, baseattr, str);
+         return;
+      }
```

Comment:
Could we try to avoid making this large case statement more convoluted and complex than necessary by factoring this into some separate function? The code inside the case statement should be as concise as possible as the cyclomatic complexity added through the switch statement is already high enough. Also make the code more easy to follow.

Reply:
Agreed